File: itraj.py
# Itraj.py
# Thin wrapper interface for QIST Python
# Author: David Cunningham
import numpy as np
from pyqist import pq

class Itraj(object):
    """
    class Itraj
    Description:
        Object-oriented wrapper for the QIST propagation runtime.
        Basic usage: initialize from a namelist file as specified in the users
        guide and then use the functions below to propagate relative motion
        from the model specified by the namelist.

        NOTE: Currently Python can only load one QIST model at a time per process, 
        so something like 
        it1 = Itraj(nml1)
        it2 = Itraj(nml2)
        will result in the statement it1==it2 evaluating to True.
    """
    def __init__(self, namelist):
        """
        Itraj initializer
        inputs: 
               namelist (string) -- a file on disk containing the ITRAJ_CONFIG namelist
        Description: Loads the specified namelist into memory for use by QIST.

        """
        pq.pw_init_n(namelist)
    def state(self,t):
        """
        Itraj.state(t)
        inputs: 
               t (float) -- time in seconds past J2000
        returns: 
               state (float)
        Description: returns the ``state'' stored in the QIST model. In the current version
            of QIST, this is just the value of an independent variable.

        """
        return pq.pw_state(t)
    def stm(self,t):
        """
        Itraj.stm(t)
        inputs: 
               t (float) -- time in seconds past J2000
        returns: 
               stm (8x8 ndarray)
        Description: returns the STM around the reference trajectory from time t0 to time t
            where t0 is the initial time specified in the namelist.
            The 8 rows/columns of the STM correspond to:
            [x y z xdot ydot zdot t0 TOF]

            If you don't need time partials then just do:
            this_stm = it.stm(t)[:6,:6]

        """
        return pq.pw_stm(t)
    def stt(self,t):
        """
        Itraj.stt(t)
        inputs: 
               t (float) -- time in seconds past J2000
        returns: 
               stt (8x8x8 ndarray)
        Description: returns the STT around the reference trajectory from time t0 to time t
            where t0 is the initial time specified in the namelist.
            The 8 pages/rows/columns of the STM correspond to:
            [x y z xdot ydot zdot t0 TOF]

            If you don't need time partials then just do:
            this_stt = it.stt(t)[:6,:6,:6]

        """
        return pq.pw_stt(t)
    def stmInv(self,t):
        """
        Itraj.stmInv(t)
        inputs: 
               t (float) -- time in seconds past J2000
        returns: 
               invstm (8x8 ndarray)
        Description: returns the inverse STM around the reference trajectory from time t to time t0
            where t0 is the initial time specified in the namelist.
        """
        return pq.pw_stm_i(t)
    def sttInv(self,t):
        """
        Itraj.sttInv(t)
        inputs: 
               t (float) -- time in seconds past J2000
        returns: 
               invstt (8x8x8 ndarray)
        Description: returns the inverse STT around the reference trajectory from time t to time t0
            where t0 is the initial time specified in the namelist.
        """
        return pq.pw_stt_i(t)
    def prop(self,ta,tb,xa,order=None):
        """
        Itraj.prop(ta,tb,xa,order=2)
        inputs: 
                ta (float) -- initial time in seconds past J2000
                tb (float) -- final time in seconds past J2000
                xa (8 x n ndarray) -- Initial relative states to propagate between times ta and tb
                                      one column should contain a vector 
                                      [dx, dy, dz, dxdot, dydot, dzdot, dt0, dTOF]
                                      to be propagated to time tb.
                                      The reference frame must be in SPICE J2000 (ICRF), 
                                      km, km/s, and s, relative to the reference trajectory.
                order (integer) -- Order of propagation, should be 1 or 2. 
                                   Other values will throw an error. 
                                   1 will cause QIST to propagate linearly, 
                                   i.e. with the STM only. 
                                   2 will cause QIST to propagate a quadratic
                                   solution, i.e. using the STM and STT.
        returns: 
                xb (8 x n ndarray) -- the final states at time tb.
        Description: The core function of QIST: propagates relative states between times.
        """
        ordr= 2
        func = {1: pq.pw_prop_once, 2: pq.pw_prop_many}
        r = xa.ndim
        if order is not None : ordr=order
        return func[r](ta,tb,xa, ordr)
    def sttsAToB(self,ta,tb):
        """
        Itraj.sttsAToB(ta,tb)
        inputs: 
                ta (float) -- initial time in seconds past J2000
                tb (float) -- final time in seconds past J2000
        returns: 
                (STMab, STTab) :: ((8x8 ndarray), (8x8x8 ndarray)) -- STM and STT from 
                                   time ta to time tb in SPICE J2000 frame in km, km/s, and s.
        Description: Returns the "chained" STM and STT between any two times in the simulation for targeting,
                     optimization, uncertainty propagation, or any other application requiring partials of the flow.
        """
        scratchm, scratcht = pq.pw_stts_ab(ta,tb)
        stm = scratchm
        stt = scratcht #np.einsum('kji -> ijk', scratcht)
        return stm,stt
    def sttUpdate(self,ta,tb,xa):
        """
        Itraj.sttUpdate(ta,tb,xa)
        inputs: 
                ta (float) -- initial time in seconds past J2000
                tb (float) -- final time in seconds past J2000
                xa (8 x n ndarray) -- Initial relative states to propagate between times ta and tb
                                      one column should contain a vector 
                                      [dx, dy, dz, dxdot, dydot, dzdot, dt0, dTOF]
                                      to be propagated to time tb.
                                      The reference frame must be in SPICE J2000 (ICRF), 
                                      km, km/s, and s, relative to the reference trajectory.
        returns: (STMab_new, STTab_new) :: ((8x8 ndarray), (8x8x8 ndarray)) -- STM and STT from 
                                   time ta to time tb in SPICE J2000 frame in km, km/s, and s.
        Description: Returns an approximation of the STM and STT of the _relative_ trajectory using the initial
                     relative state and the _reference_ trajectory STM and STT.
        """
        stm, stt = pq.pw_stt_update(ta,tb,xa)
        return stm,stt
    def tensorChangeBasis(self,RNewOldf, ROldNew0, old_stm, old_stt):
        """
        Itraj.tensorChangeBasis(RNewOldf, ROldNew0, old_stm, old_stt)
        inputs: 
                RNewOldf (8 x 8 ndarray) -- Transformation matrix from the old basis to the new basis at time tf
                ROldNew0 (8 x 8 ndarray) -- Transformation matrix from the new basis to the old basis at time t0
                old_stm  (8 x 8 ndarray) -- STM from time ta to tb in old basis
                old_stt  (8 x 8 x 8 ndarray) -- STT from time ta to tb in old basis
        returns: (STMab_new, STTab_new) :: ((8x8 ndarray), (8x8x8 ndarray)) -- STM and STT from 
                                   time ta to time tb in km, km/s, and s in new basis.
        Description: This function applies a rotation matrix or other linear transformation to the STM and STT
                     to change their basis. Useful for when a user wants to propagate or otherwise operate 
                     in a frame other than J2000.
                    
        """
        return pq.pw_tensor_change_of_basis(RNewOldf, ROldNew0, old_stm, old_stt)

File: itraj.m
from addlib import *
import numpy as np
from pyqist import pq

class Itraj(object):
    """
    A lightweight wrapper class for running QIST models in Python.
    

    Parameters
    ----------
    None

    Examples
    --------
    >>> from itraj import Itraj
    >>> it = Itraj(path="./examples/",nml="curve_deimos_config_namelist_2026Nov26120000002026Nov2617450000.nml")
    """
    def __init__(self, namelist):
        """
        namelist             : character
        """
        pq.pw_init_n(namelist)
    def state(self,t):
        """
        Returns the reference state at a time t.
        For this version of QIST, the state is just the time in seconds
        past J2000, since the reference state is never computed by QIST
        """
        return pq.pw_state(t)
    def stm(self,t):
        """
        Returns the STM at a time t in seconds past J2000
        """
        return pq.pw_stm(t)
    def stt(self,t):
        """
        Returns the STT at a time t in seconds past J2000
        """
        return pq.pw_stt(t)
    def stmInv(self,t):
        """
        Returns the inverse STM at a time t in seconds past J2000
        """
        return pq.pw_stm_i(t)
    def sttInv(self,t):
        """
        Returns the inverse STT at a time t in seconds past J2000
        """
        return pq.pw_stt_i(t)
    def prop(self,ta,tb,xa,order=None):
        """
        Propagates relative states.
        
        Inputs
        ------
        ta: float, time in seconds past J2000 TDB to start propagation
        tb: float, time in seconds past J2000 TDB to end propagation
        xa: float, dimension (8,:), initial relative state in km and km/s with two padding zeros

        Returns the state at a time tb in seconds past J2000 with two padding zeros
        """
        ordr= 2
        func = {1: pq.pw_prop_once, 2: pq.pw_prop_many}
        r = xa.ndim
        if order is not None : ordr=order
        return func[r](ta,tb,xa, ordr)
    def sttsAToB(self,ta,tb):
        """
        Gets stm and stt from a time ta to a time tb
        
        Inputs
        ------
        ta: float, time in seconds past J2000 TDB to start propagation
        tb: float, time in seconds past J2000 TDB to end propagation

        Returns
        ------
        stm: float (8,8), first order STM of the reference trajectory
        stt: float (8,8,8), second order STT of the reference trajectory
        """
        stm, stt = pq.pw_stts_ab(ta,tb)
        return stm,stt
    def sttUpdate(self,ta,tb,xa):
        """
        Returns approximate STM and STT of the chaser with state xa from time ta to time tb
        Inputs
        ------
        ta: float, time in seconds past J2000 TDB to start propagation
        tb: float, time in seconds past J2000 TDB to end propagation
        xa: float, dimension (8,:), initial relative state in km and km/s with two padding zeros

        Returns
        ------
        stm: float (8,8), approximate first order STM of the CHASER trajectory
        stt: float (8,8,8), approximate second order STT of the CHASER trajectory
        """
        stm, stt = pq.pw_stt_update(ta,tb,xa)
        return stm,stt
    def tensorChangeBasis(self,RNewOldf, ROldNew0, old_stm, old_stt):
        """
        Transform STM and STT from an old basis to a new basis. Rotation matrix epochs
        should correspond to the times at which the original STM and STT were evaluated.
        Inputs
        ------
        RNewOldf: float, (8,8) Matrix with rotation matrix in the upper left 6x6 block
                               to transform from the new frame to the old frame at time tf
        ROldNew0: float, (8,8) Matrix with rotation matrix in the upper left 6x6 block
                               to transform from the old frame to the new frame at time t0
        old_stm: float, dimension (8,8), STM in old coordinate frame from time t0 to tf
        old_stt: float, dimension (8,8,8), STT in old coordinate frame from time t0 to tf

        Returns
        ------
        new_stm: float (8,8),  STM in new coordinate frame from time t0 to tf
        new_stt: float (8,8,8),  STT in new coordinate frame from time t0 to tf
        """
        return pq.pw_tensor_change_of_basis(RNewOldf, ROldNew0, old_stm, old_stt)
    def zMap(self,t,order):
        """
        Propagates a test state to a time after the model epoch and then back to the
        origin. The result should be zero.

        Inputs
        ------
        t: float, in seconds past J2000 TDB
        order: 1 or 2

        Returns
        ------
        some numbers very close to zero
        """
        return pq.pw_zmap(t,order)
